frontend/lib/resolution/Resolver.h

```diff
@@ -475,37 +470,103 @@ struct Resolver {
+                         std::vector<const uast::VarLikeDecl*>&)>;
+
+    static void reportNoMatchingCandidates(const ResolvedCallResult& r,
+                                           std::vector<ApplicabilityResult>& rejected,
+                                           std::vector<const uast::VarLikeDecl*>& actualDecls) {
+      CHPL_REPORT(r.parent->context, NoMatchingCandidates,
+                  r.astForContext, *r.ci, rejected, actualDecls);
+    }
+
+    Resolver* parent = nullptr;
+    CallResolutionResult result;
+    uast::AstNode const* astForContext = nullptr;
+    CallInfo const* ci = nullptr;
+    CallScopeInfo const* inScopes = nullptr;
+    types::QualifiedType receiverType;
+    bool wasGeneratedCall = true;
+    std::vector<const uast::AstNode*> const* actualAsts = nullptr;
+    const char* callName = nullptr;
+    ReportFn reportError = reportNoMatchingCandidates;
+
+    // issue ambiguity / no matching candidates / etc error
+    void issueBasicError();
+
+    // handle the result of one of the functions to resolve a call. Handles:
+    //  * r.setMostSpecific
+    //  * r.setPoiScope
+    //  * r.setType
+    //  * poiInfo.accumulate
+    //  * userDiagnostics
+    //
+    // Does not handle:
+    //
+    //  * issueErrorForFailedCallResolution if there was an error
+    //
+    // Instead, returns 'true' if an error needs to be issued.
+    bool noteResultWithoutError(ResolvedExpression* r,
+                                optional<ActionAndId> associatedActionAndId = {});
+
+    static bool noteResultWithoutError(Resolver& resolver,
+                                       ResolvedExpression* r,
+                                       const uast::AstNode* astForContext,
+                                       const CallResolutionResult& c,
+                                       optional<ActionAndId> associatedActionAndId = {});
+
+    // Same as noteResultWithoutError, but also issues errors.
+    void noteResult(ResolvedExpression* r,
+                    optional<ActionAndId> associatedActionAndId = {});
+
+    // Issues a more specific error (listing rejected candidates) if possible.
+    // To collect the candidates, re-runs the call. Returns true if an error
+    // was emitted.
+    bool rerunCallAndPrintCandidates();
+
+    // Like noteResult, except attempts to do more work to print fancier errors
+    // (see rerunCallAndPrintCandidates).
+    void noteResultPrintCandidates(ResolvedExpression* r,
+                                   optional<ActionAndId> associatedActionAndId = {});
+  };
+
+  /* The resolver's wrapper of resolution::resolveGeneratedCall.
+     Stores additional information into a ResolvedCallResult to enable
+     updating the resolver with results of the call if needed.
+   */
+  ResolvedCallResult resolveGeneratedCall(const uast::AstNode* astForContext,
+                                          const CallInfo* ci,
+                                          const CallScopeInfo* inScopes,
```

Comment:
Do ``CallInfo`` and ``CallScopeInfo`` need to be pointers here?

Reply:
I had three options here: to pass by (r)value, pointer, or reference. The trouble is that the `ResolvedCallResult` needs to have a default constructor, because it's used as a result from `runAndTrackErrors` (which requires the type it returns to have a default init). This ruled out references (can't default-initialize a reference), as well as (r)values, because `CallResolutionResult` doesn't have a default constructor. I also observed that neither the `CallInfo` nor the `CallScopeInfo` outlive the `ResolvedCallResult` in `Resolver.cpp`, so it made sense to pass them as pointers, which have a nice default-init value of `nullptr` and in fact take less space than copies of `ci` and `inScopes`.